book/src/specification/logical.md

```diff
@@ -1,3 +1,788 @@
+ - \\(F_{signals} := \textrm{fields}(T_{signals})\\)
+ - For all \\(i \in (1, 2, ..., n)\\):
+    - Unpack \\(T_i\\) into \\(\textrm{Stream}(T_e, n, d, s, c, r, T_u, x)\\).
+    - \\(P_i := \textrm{PhysicalStream}(\textrm{fields}(T_e), \lceil n\rceil, d, c, \textrm{fields}(T_u))\\)
+ - Return \\(\textrm{Physical}(F_{signals}, N_1 : P_1, N_2 : P_2, ..., N_n : P_n)\\).
+
+#### Type compatibility function
+
+This section defines the function
+\\(\textrm{compatible}(T_{source}, T_{sink}) \rightarrow \textrm{Bool}\\),
+where \\(T_{source}\\) and \\(T_{sink}\\) are both logical stream types. The
+function returns true if and only if a source of type \\(T_{source}\\) can be
+connected to a sink of type \\(T_{sink}\\) without insertion of conversion
+logic. The function returns true if and only if one or more of the following
+rules match:
+
+ - \\(T_{source} = T_{sink}\\);
+
+ - \\(T_{source} = \textrm{Stream}(T_e, n, d, s, c, r, T_u, x)\\),
+   \\(T_{sink} = \textrm{Stream}(T'_d, n, d, s, c', r, T_u, x)\\),
+   \\(\textrm{compatible}(T_e, T'_d) = \textrm{true}\\), and \\(c < c'\\);
+
+ - \\(T_{source} = \textrm{Group}(N_1: T_1, N_2: T_2, ..., N_n: T_n)\\),
+   \\(T_{sink} = \textrm{Group}(N_1: T'_1, N_2: T'_2, ..., N_n: T'_n)\\),
+   and \\(\textrm{compatible}(T_i, T'_i)\\) is true for all
+   \\(i \in (1, 2, ..., n)\\); or
+
+ - \\(T_{source} = \textrm{Union}(N_1: T_1, N_2: T_2, ..., N_n: T_n)\\),
+   \\(T_{sink} = \textrm{Union}(N_1: T'_1, N_2: T'_2, ..., N_n: T'_n)\\),
+   and \\(\textrm{compatible}(T_i, T'_i)\\) is true for all
+   \\(i \in (1, 2, ..., n)\\).
+
+The name matching is to be done case sensitively.
+
+> While two streams may be compatible in a case-insensitive language when only
+> the case of one or more of the names differ, this would still not hold for
+> case-sensitive languages. Therefore, for two streams to be compatible in all
+> cases, the matching must be case sensitive.
+
+Union semantics
+---------------
+
+When flattened into fields, a \\(\textrm{Union}\\) node consists of:
+
+ - if there are two or more variants, an `"tag"` field of size
+   \\(\left\lceil\log_2{n}\right\rceil\\), where \\(n\\) is the number of
+   variants; and
+ - if any variant carries data in the same stream as the union itself, a
+   `"union"` field of size\
+   \\(\max\limits_{\textrm{variants}} \sum\limits_{\textrm{subfields}} b_{\textrm{subfield}}\\).
+
+The `"tag"` field is used to represent which variant is being represented, by
+means of a zero-based index encoded as an unsigned binary number.
+
+It is illegal for a source to drive the `"tag"` field of a \\(\textrm{Union}\\)
+with \\(n\\) variants to \\(n\\) or higher.
+
+> The above would otherwise be possible if \\(n\\) is not a power of two.
+
+The `"union"` field of a \\(\textrm{Union}\\) is used to represent variant
+data existing in the same stream as the union itself. It consists of the
+LSB-first, LSB-aligned concatenation of the fields of the variant.
```

Comment:
why concatenation?

Reply:
It's talking about the fields of a single variant, i.e. if you have `Union(a: Group(x:..., y:...), b:...)` then `a` will have the fields `x` and `y`. These are concatenated together into a single "field", so it can be overlaid with `a`.